### PR TITLE
Refactor the protocol and handshake

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

File: test/signalrclienttests/websocket_transport_tests.cpp [hunk 1]
```diff
@@ -394,19 +362,20 @@ TEST(websocket_transport_disconnect, disconnect_is_no_op_if_transport_not_starte
     ASSERT_FALSE(close_called);
 }
 
-TEST(websocket_transport_disconnect, exceptions_from_outstanding_receive_task_observed_after_websocket_transport_disconnected)
+// This is an odd test
+TEST(websocket_transport_disconnect, DISABLED_exceptions_from_outstanding_receive_task_observed_after_websocket_transport_disconnected)
```
Comment (halter73) [inline test/signalrclienttests/websocket_transport_tests.cpp]: So why is this disabled?

File: test/signalrclienttests/websocket_transport_tests.cpp [hunk 2]
```diff
@@ -478,10 +423,12 @@ void receive_loop_logs_exception_runner(const T& e, const std::string& expected_
...
     mre.get();
 
-    receive_event.wait();
+    ASSERT_FALSE(client->receive_loop_started.wait(5000));
+
+    client->receive_message(std::make_exception_ptr(e));
 
     // this is race'y but there is nothing we can block on
-    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (halter73) [inline test/signalrclienttests/websocket_transport_tests.cpp]: Can we check the following assertion in a loop until it succeeds or fails enough retries to make this less racy? Or can we use a test log writer that trips a cancellation_token once the expected log is logged?

Comment (BrennanConroy) [reply test/signalrclienttests/websocket_transport_tests.cpp]: Sure, I'll attempt to make it a little better

File: test/signalrclienttests/websocket_transport_tests.cpp (hunk 1 above)
Comment (BrennanConroy) [reply test/signalrclienttests/websocket_transport_tests.cpp]: We might want to delete the test. I think it was checking that a crash didn't happen back when we used `pptx::task`, as that would crash for unobserved exceptions in tasks.

Or we might want to refactor this to check that an exception is logged/bubbled up.

**Pull request merged**